[Infra] Create new formulas for rx and tx metrics

jennypavlova opened the pull request:
Closes #188641

## Summary

This PR adds new formulas for rx and tx metrics for hosts. In inventory we show the old metrics as legacy and the new ones with the old metrics labels (this affects only hosts):

<img width="1788" alt="image" src="https://github.com/user-attachments/assets/d3e5bf26-e521-4ff8-b00b-1d78eebd56f9">

All old alerts should work - The only difference is that it will show the metric as "Legacy" and it still can be used in the rules. The hosts view and the lens charts are using a new formula

## Testing
- Check the network metrics in the inventory / alert flyout (both the new ones and the old ones)
- Check the network metrics and charts in the hosts view (only the new ones should be available)

https://github.com/user-attachments/assets/886fd5a0-858c-458b-9025-eb55913b1932


https://github.com/user-attachments/assets/7752939f-f693-4021-bf23-89e264ef0c2d

jennypavlova committed:
Update lens rx and tx charts

jennypavlova committed:
add new RX and TX values

jennypavlova committed:
Change rx/tx queries

jennypavlova committed:
Add condition for legacy labels

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

jennypavlova commented on x-pack/plugins/observability_solution/metrics_data_access/common/inventory_models/host/metrics/snapshot/rx_new.ts:
```diff
@@ -0,0 +1,39 @@
+ */
+
+import { MetricsUIAggregation } from '../../../types';
+export const rxNew: MetricsUIAggregation = {
```
I know it will be better to rename the old ones to legacy instead but this will affect the existing alerts (see https://github.com/elastic/kibana/issues/188641#issuecomment-2248451883)

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

crespocarlos commented on x-pack/plugins/observability_solution/metrics_data_access/common/inventory_models/types.ts:
```diff
@@ -358,6 +358,8 @@ export const SnapshotMetricTypeKeys = {
   normalizedLoad1m: null,
   tx: null,
   rx: null,
+  txNew: null,
+  rxNew: null,
```
Could we name  `rxNew` or `txNew` differently? I'm not sure what's better, but perhaps `tx_v2` or `rx_v2`, or `rx_rate_per_sec` or `tx_rate_per_sec`?

jennypavlova replied:
Thanks, I used V2 as it will be easier to add new versions after in case we need to.

crespocarlos reviewed (commented):
I had a quick look at it, I'll test it tomorrow. It looks good, I'm just curious if we could find a better name for the new formulas

crespocarlos reviewed (commented):
I've tested it and it works. Alerts created with the legacy rules continue to work after starting this branch.

A few things though:

Instead of `(sum.)` it should be `(avg.)`. I know we have a logic to extract the formula and display it next to the label, but this time `sum` is wrong.
<img width="500px" alt="image" src="https://github.com/user-attachments/assets/6a8fa45e-c5f0-4063-9ebd-b29e80da5c6e">


@roshan-elastic CPU label [will not be `legacy` anymore](https://github.com/elastic/kibana/issues/188634#issuecomment-2248354019), we should do the same with RX and TX. How about  "Inbound/Outbount Traffic (per metricset.period)" and "Inbound/Outbount Traffic (per second)"? - Sorry @jennypavlova, I forgot to follow up and didn't update the AC.
<img width="500px" alt="image" src="https://github.com/user-attachments/assets/17f7df97-f7d8-47b0-88ca-2a24a8d33f86">

roshan-elastic commented:
> @roshan-elastic CPU label [will not be `legacy` anymore](https://github.com/elastic/kibana/issues/188634#issuecomment-2248354019), we should do the same with RX and TX. How about "Inbound/Outbount Traffic (per metricset.period)" and "Inbound/Outbount Traffic (per second)"? - Sorry @jennypavlova, I forgot to follow up and didn't update the AC. <img alt="image" width="500px" src="https://private-user-images.githubusercontent.com/2767137/353334719-17f7df97-f7d8-47b0-88ca-2a24a8d33f86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzMjcyNDcsIm5iZiI6MTcyMjMyNjk0NywicGF0aCI6Ii8yNzY3MTM3LzM1MzMzNDcxOS0xN2Y3ZGY5Ny1mN2Q4LTQ3YjAtODhjYS0yYTI0YThkMzNmODYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDczMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MzBUMDgwOTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmZkMjhhMDRhMWY4Y2JiOWUyM2NiZTMwMzMwZmU3ZTM2M2VmY2FkODgyZmNlODBlMjA5ZDY5ZmVmYTQyNmM2OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ShOpJDZzi7MJsuDz9fBvWjFZ_MY8PUACC55RHYy8tHE">

Hey @crespocarlos, we could do that but I have a concern that unlike the CPU metrics, where both CPU calculations **might** be useful, the legacy metrics for RX/TX would never be chosen by users (and we should not indicate it is a viable option for users to choose).

In this case, I would prefer to leave this as a 'legacy' so users can start moving over to use the new one. My concern here is that if we force the user to try and understand the difference - they'll just get confused (e.g. "what is a `metricset.period`?").

In which case, I think we should be opinionated on this one and basically tell the users it is 'legacy' and subtly encourage them to move to the new one (by not giving both the new and old equal weight in the UI).

If we give explanations of the difference in the UI, we can let them understand why we're calling it 'legacy'.

Just so I understand the new vs old behaviour though:

**New** : Shows the value per 'second' (no matter the time range)
**Old** : This rate was based on the `metricset.period` defined in their agent/beats

On the 'old' one, would the rate change based on the period selected or would it always be a consistent rate?

jennypavlova committed:
CR changes

jennypavlova committed:
Fix tests

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

wayneseymour reviewed (approved):
LGTM, cr only.

jennypavlova commented:
@crespocarlos Thank you for reviewing and testing the PR!

> Instead of (sum.) it should be (avg.). I know we have a logic to extract the formula and display it next to the label, but this time sum is wrong.

Good point, fixed ✅ 

> "Inbound/Outbount Traffic (per metricset.period)" and "Inbound/Outbount Traffic (per second)"?

Just checking this here - we don't use per second inside the rule now because of the changes and now I set it to be just bits - should I update something there 🤔 

<img width="200" alt="image" src="https://github.com/user-attachments/assets/13da2b60-093a-4fac-9ad4-167443b7b1e6"> <img width="200" alt="image" src="https://github.com/user-attachments/assets/fde0ddf7-d69a-44aa-a250-0c381912c64a">

> In this case, I would prefer to leave this as a 'legacy' so users can start moving over to use the new one. My concern here is that if we force the user to try and understand the difference - they'll just get confused (e.g. "what is a metricset.period?").

This makes sense - so we want to keep `... (Legacy)` label and still show old and new right? (I mean just for `Hosts` as the other assets don't have the Legacy because they don't have new formulas)

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

crespocarlos commented on x-pack/test/api_integration/apis/metrics_ui/infra.ts:
```diff
@@ -247,7 +247,7 @@ export default function ({ getService }: FtrProviderContext) {
 
         expect(normalizeNewLine(response.body.message)).to.be(
-          '[request body]: Failed to validate: in metrics/0/type: "any" does not match expected type "cpu" | "normalizedLoad1m" | "diskSpaceUsage" | "memory" | "memoryFree" | "rx" | "tx"'
+          '[request body]: Failed to validate: in metrics/0/type: "any" does not match expected type "cpu" | "normalizedLoad1m" | "diskSpaceUsage" | "memory" | "memoryFree" | "rx" | "tx" | "rxNew" | "txNew"'
```
```suggestion
          '[request body]: Failed to validate: in metrics/0/type: "any" does not match expected type "cpu" | "normalizedLoad1m" | "diskSpaceUsage" | "memory" | "memoryFree" | "rx" | "tx" | "rxV2" | "txV2"'
```

crespocarlos commented on x-pack/plugins/observability_solution/infra/public/pages/metrics/inventory_view/lib/create_inventory_metric_formatter.ts:
```diff
@@ -65,6 +65,8 @@ const METRIC_FORMATTERS: MetricFormatters = {
   },
   rx: { formatter: InfraFormatterType.bits, template: '{{value}}/s' },
   tx: { formatter: InfraFormatterType.bits, template: '{{value}}/s' },
+  rxV2: { formatter: InfraFormatterType.bits, template: '{{value}}' },
+  txV2: { formatter: InfraFormatterType.bits, template: '{{value}}' },
```
What happens if you add the `/s` here?
```suggestion
  rxV2: { formatter: InfraFormatterType.bits, template: '{{value}}/s' },
  txV2: { formatter: InfraFormatterType.bits, template: '{{value}}/s' },
```

crespocarlos commented on x-pack/plugins/observability_solution/infra/public/alerting/inventory/components/expression.tsx:
```diff
@@ -775,6 +775,8 @@ const metricUnit: Record<string, { label: string }> = {
   memory: { label: '%' },
   rx: { label: 'bits/s' },
   tx: { label: 'bits/s' },
+  rxV2: { label: 'bits' },
+  txV2: { label: 'bits' },
```
Here should be 
```suggestion
  rxV2: { label: 'bits/s' },
  txV2: { label: 'bits/s' },
```

crespocarlos reviewed (commented):
> Just checking this here - we don't use per second inside the rule and inventory now because of the changes and now I set it to be just bits - should I update something there 🤔

hey @jennypavlova , good catch. I think we need to have the `/s` there too

jennypavlova committed:
CR, tooltip metrics and test fixes

jennypavlova committed:
Merge branch '188641-infra-create-new-formulas-for-rx-and-tx-metrics' of https://github.com/jennypavlova/kibana into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

jennypavlova commented:
@crespocarlos Thanks I fixed the format to include per second and thanks to the test I found an issue with the tooltip metrics (still showing the `Legacy` ones) so I changed that as well: 
<img width="1561" alt="image" src="https://github.com/user-attachments/assets/6d45793c-8f1b-4a9d-8120-7de7bb855e58">

crespocarlos reviewed (approved):
LGTM. Thanks!

jennypavlova committed:
Fix format

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

jennypavlova committed:
Fix type

jennypavlova committed:
Merge branch '188641-infra-create-new-formulas-for-rx-and-tx-metrics' of https://github.com/jennypavlova/kibana into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

jennypavlova committed:
Fix type lint

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

jennypavlova committed:
fix lint

jennypavlova committed:
Merge branch '188641-infra-create-new-formulas-for-rx-and-tx-metrics' of https://github.com/jennypavlova/kibana into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

crespocarlos committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

jennypavlova commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

jennypavlova committed:
Test workaround

jennypavlova commented on x-pack/test/functional/page_objects/infra_hosts_view.ts:
```diff
@@ -129,8 +129,6 @@ export function InfraHostsViewProvider({ getService }: FtrProviderContext) {
       await button.click();
       await testSubjects.existOrFail('embeddablePanelAction-openInLens');
-      // forces the modal to close
-      await element.click();
```
I think somehow this click opened the modal and this test failed multiple times:
<img width="1527" alt="image" src="https://github.com/user-attachments/assets/5781db4c-f05f-4c87-bc15-e9411954c47b">


I don't know why it happens only on this PR - my assumption is that with a different TX formula now it clicks on the chart line 🤔 

In the end, I used an escape button key press to close it in [hosts_view.ts](https://github.com/elastic/kibana/pull/189281/files#diff-cdf2ef63d7f0db051d569b09fbd2ba39eb852d3d75bf3b29d85e57de2da3af35) to avoid unexpected behavior

jennypavlova committed:
Merge branch 'main' into 188641-infra-create-new-formulas-for-rx-and-tx-metrics

kdelemme reviewed (approved):
ux-management code change LGTM!
